src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -303,64 +311,57 @@ void PipelineExecutor::initializeExecution(size_t num_threads)
+    pool.setMaxThreads(num_threads);
 }
 
 void PipelineExecutor::spawnThreads()
 {
     while (auto slot = slots->tryAcquire())
     {
-        std::unique_lock lock{threads_mutex};
-        size_t thread_num = threads.size();
+        size_t thread_num = threads++;
 
         /// Count of threads in use should be updated for proper finish() condition.
         /// NOTE: this will not decrease `use_threads` below initially granted count
         tasks.upscale(thread_num + 1);
 
         /// Start new thread
-        threads.emplace_back([this, thread_num, thread_group = CurrentThread::getGroup(), slot = std::move(slot)]
+        try
         {
-            /// ThreadStatus thread_status;
-
-            SCOPE_EXIT_SAFE(
-                if (thread_group)
-                    CurrentThread::detachFromGroupIfNotDetached();
-            );
-            setThreadName("QueryPipelineEx");
-
-            if (thread_group)
-                CurrentThread::attachToGroup(thread_group);
-
-            try
-            {
-                executeSingleThread(thread_num);
-            }
-            catch (...)
+            pool.scheduleOrThrowOnError([this, thread_num, thread_group = CurrentThread::getGroup(), slot = std::move(slot)]
             {
-                /// In case of exception from executor itself, stop other threads.
-                finish();
-                tasks.getThreadContext(thread_num).setException(std::current_exception());
-            }
-        });
-    }
-}
+                /// ThreadStatus thread_status;
 
-void PipelineExecutor::joinThreads()
-{
-    for (size_t thread_num = 0; ; thread_num++)
-    {
-        std::unique_lock lock{threads_mutex};
-        if (thread_num >= threads.size())
-            break;
-        if (threads[thread_num].joinable())
+                SCOPE_EXIT_SAFE(
+                    if (thread_group)
+                        CurrentThread::detachFromGroupIfNotDetached();
+                );
+                setThreadName("QueryPipelineEx");
+
+                if (thread_group)
+                    CurrentThread::attachToGroup(thread_group);
+
+                try
+                {
+                    executeSingleThread(thread_num);
+                }
+                catch (...)
+                {
+                    /// In case of exception from executor itself, stop other threads.
+                    finish();
+                    tasks.getThreadContext(thread_num).setException(std::current_exception());
+                }
+            });
+        }
+        catch (...)
         {
-            auto & thread = threads[thread_num];
-            lock.unlock(); // to avoid deadlock if thread we are going to join starts spawning threads
-            thread.join();
+            /// In case of exception from executor itself, stop other threads.
+            finish();
+            tasks.getThreadContext(thread_num).setException(std::current_exception());
+
+            pool.wait();
```

Comment:
As I understand, this is a deadlock.
`(spawnThreads) pool.scheduleOrThrowOnError -> executeSingleThread -> executeStepImpl -> spawnThreads -> pool.join()`

`pool.wait()` inside `PipelineExecutor::executeImpl` should be enough, I suppose.

Reply:
It should not.

`executeSingleThread` will not throw, it will save the exception and rethrow it later, and this `pool.wait()` is mostly for cases when thread allocation fails.

Reply:
But indeed pool.wait looks redundant anyway - removed.